### PR TITLE
Add Janosch's existing slab solver code

### LINKED ISSUE #4: Pull out slabsolver core from Janosch's repo.

### LINKED ISSUE #6: Minimal refactoring to make it import-able

### PR DESCRIPTION
**Pull request by matt-graham**
Resolves #4 and resolves #6.

Pulls in @janoschpreuss's existing slab solver implementation, adds all directly imported packages as dependencies, and does a very minimal refactoring (renaming modules to be in snake_case and using package based imports) to make it importable as a package.

The commits here were with `--no-verify` to skip the pre-commit linting / formatting and linting checks will fail. We could potentially run `black` and `isort` on the code to at least deal with formatting issues, but I've not done this so far to keep the changes minimal - @samcunliffe and @krishnakumarg1984 any thoughts on this?

### CODE REVIEW

**Comment by samcunliffe** [comment]
You're fast!

> but I've not done this so far to keep the changes minimal - @samcunliffe and @krishnakumarg1984 any thoughts on this?

I tend to agree. At the very least this way lets Janosch have a look at his own code to check everything is OK with the packaging. That is: nothing omitted and nothing erroneous included.

I'd say do linting in it's own PR (which will be very noisy).

------

Only the dummy test is running. Did you mean to also run the `tests/slab_solver_test_*d.py`? Should be a simple rename to have `test_` first.

File: pyproject.toml [hunk 1]
```diff
@@ -20,7 +20,14 @@ classifiers = [
 dependencies = [
     "fenics-dolfinx>=0.6.0",
     "fenics-ufl",
+    "h5py",
+    "matplotlib",
+    "mpi4py",
     "numpy",
+    "petsc4py",
+    "pypardiso",
```
Comment (samcunliffe) [inline pyproject.toml:28]: `pypardiso` interfaces to Intel MKL ... I have an M2. This may be a headache.

Comment (matt-graham) [reply pyproject.toml:28]: We can probably remove the `pypardiso` dependency or at least make optional as I think this was just a result of Janosch trying out options for shared-memory parallelised solves after I mentioned the `pypardiso` might be an option.

Comment (samcunliffe) [reply pyproject.toml:28]: An optional dep was my first thought. He already has the if statements there so can `try` the import or something. (Whatever is least ugly.) Obviously removing it completely would also make me happy 🤣.

* https://github.com/haasad/PyPardisoProject/issues/44

File: pyproject.toml [hunk 2]
```diff
@@ -20,7 +20,14 @@ classifiers = [
 dependencies = [
     "fenics-dolfinx>=0.6.0",
     "fenics-ufl",
```
Comment (samcunliffe) [inline pyproject.toml:22]:
```suggestion
    "fenics-ufl",
    "gmsh",
```
?

File: pyproject.toml [hunk 3]
```diff
@@ -20,7 +20,14 @@ classifiers = [
...
+    "h5py",
```
Comment (samcunliffe) [inline pyproject.toml:23]:
```suggestion
    "gmsh",
    "h5py",
```

File: pyproject.toml [hunk 4]
```diff
@@ -159,9 +166,16 @@ python =
 
 [testenv]
 conda_deps =
-    fenics-dolfinx
+    fenics-dolfinx>=0.6.0
     fenics-ufl
+    h5py
```
Comment (samcunliffe) [inline pyproject.toml:171]:
```suggestion
    gmsh
    h5py
```

**Review by janoschpreuss — APPROVED** [review]
I am ok with the code. However, it would be nice to keep compatability to later versions of `dolfinx`, see branch https://github.com/UCL/dxss/tree/fenics_dev

**Comment by samcunliffe** [comment]
Yeah. We chatted last week. Our (or at least my) aim is to get a build of your working commit against which we can also run the tests. Ideally this will be included in the CI. We'll track in https://github.com/UCL/dxh/issues/11

If all of that works. we can also run in the CI here.

**Pull request merged**